pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -125,16 +130,15 @@ public boolean isFinal() {
   private final Schema _schema;
   private final ServerMetrics _serverMetrics;
   private final RealtimeSegmentImpl _realtimeSegment;
-  private long _endOffset = Long.MAX_VALUE; // No upper limit on Kafka offset
-  private long _currentOffset;
+  private volatile long _currentOffset;
   private volatile State _state;
-  private int _numRowsConsumed = 0;
+  private volatile int _numRowsConsumed = 0;
```

Comment:
Agreed, for this particular variable. But I thought it is better in case we want to add the num rows into a log (in cases where we consumed thousands of rows and still could not get to the final offset in the last sprint).

No, the last catchup and main thread do not run concurrently. The last catchup stops the main thread, ensures it is stopped, and then attempts to consume for a short period if needed.